core/src/main/java/com/linecorp/armeria/common/PropagatingContextAwareExecutorService.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import java.util.concurrent.ExecutorService;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.MoreObjects;
+
+final class PropagatingContextAwareExecutorService extends AbstractContextAwareExecutorService {
+    PropagatingContextAwareExecutorService(ExecutorService executor) {
+        super(executor);
+    }
+
+    @Override
+    @Nullable
+    public RequestContext context() {
+        return RequestContext.currentOrNull();
+    }
```

Comment:
`ContextHolder`, the interface this class implements, has `context()` method whose contract is to return non-null. How about making this method return `RequestContext.current()`, so it raises an exception when the context is unavailable? We could instead rename `AbstractContextAwareExecutorService.context()` to `contextOrNull()`?

Reply:
I deliberately made this class not implement `ContextHolder` to avoid that problem. I think we need to use `currentOrNull` because the same executor might be used for non-request tasks. But happy to rename this method for clarity.

Reply:
> I think we need to use currentOrNull because the same executor might be used for non-request tasks.

Shouldn't we let users use another executor for non-request tasks?
The problem is that if we allow returning `null` here, a user could not notice whether the context is propagating or not.
I think we should leave a log at least.

Reply:
Any thought on this? cc @trustin, @ikhoon

Reply:
> The problem is that if we allow returning `null` here, a user could not notice whether the context is propagating or not.

That's a good point.

How about adding an overloaded method that accepts `boolean rejectOnMissingContext` parameter? If `true` is specified and there's no context available, we could raise a `RejectedExecutionException("context not available")`. If no parameter is given, we could keep the current behavior.

Reply:
That sounds like a good idea. 👍

Reply:
The primary use case for this whole PR is an executor that's not completely under the user's control (e.g. used in the internals of some third-party library). I think in such cases it's likely that there will be non-request tasks on the executor as well. At the very least we shouldn't outright error in such cases.

I could add code to log a warning the first time a task is executed from a place with no request context (similar to what we do with Brave trace contexts)? I think more than that is likely to be too noisy.

Reply:
That also sounds good to me too.